### ISSUE TITLE
Match statement union not a class and cant use itself in its own definition.

### ISSUE THREAD

**Issue by nathan-chappell** [microsoft/pyright#6735]
**Describe the bug**
So I have a union type `AorB`, and these classes need to be added to a list, I decided to use a decorator (see code).
It seems like using the decorators in this way defeats pyright once it gets to a `match` statement.  I get two different seemingly erroneous error messages:

* `"AorB" is not a class`
* `Type alias "AorB" cannot use itself in its definition`

The second error message appears if I comment precisely one decorator below.

**Code or Screenshots**
```py
from typing import Union, TypeVar
type_list: list[type["AorB"]] = []

def add_type_to_list(t: type["AorB"]) -> type["AorB"]:
    type_list.append(t)
    return t

@add_type_to_list  # comment this line to get the other error message
class A: ...

@add_type_to_list
class B: ...

AorB = A | B  # Type alias "AorB" cannot use itself in its definition

match 0:
    case A(): ...  # "AorB" is not a class
    case B(): ...  # "AorB" is not a class
```

![image](https://github.com/microsoft/pyright/assets/36384302/d09e7755-d5c4-4c2d-b9e6-32daf6bfc69d)
![image](https://github.com/microsoft/pyright/assets/36384302/9921b6eb-24cc-4533-9ab3-a4b315d537ae)

**VS Code extension or command-line**
Extension: `Pylance language server 2023.12.1 (pyright a78e2b10) starting`

**Comment by erictraut** [comment]
This is a problem with the way you've annotated your decorator. It shouldn't use a union. Instead, it should be generic.

Using PEP 695 syntax (if you're using Python 3.12):

```python
def add_type_to_list[T: "A | B"](t: type[T]) -> type[T]:
    type_list.append(t)
    return t
```

Or if you're using an older version of Python:

```python
T = TypeVar("T", bound="A | B")

def add_type_to_list(t: type[T]) -> type[T]:
    type_list.append(t)
    return t
```

**Comment by nathan-chappell** [comment]
@erictraut can't believe I missed that, sorry!

**Comment by nathan-chappell** [comment]
@erictraut wait, I rechecked my code and something funny is going on here.  I was using generics before, then probably switched to the union-type approach above, which you correctly identified as being incorrect.

Here is the code I would like to use (sorry for the ugly names):

```py
from typing import TypeAlias, TypeVar

TAorB = TypeVar("TAorB", bound="AorB")
type_list_AorB: list[type["AorB"]] = []

def append_TAorB(t: type[TAorB]) -> type[TAorB]:
    type_list_AorB.append(t) # I get a type error here
    return t

@append_TAorB
class A:
    x: int

@append_TAorB
class B:
    x: str

AorB: TypeAlias = A | B
```

It's very close to what you suggested, but I've used an alias `AorB` instead of the literal `A | B`.  The screen capture below shows an example of how this behavior is confusing:

![pyright-typing-issues](https://github.com/microsoft/pyright/assets/36384302/b518bab9-b413-44dd-87b0-4d04005c3eef)

Trying to reason about the math, I guess if a type is bounded above by `A | B`, that doesn't necessarily mean that it's bounded above by `A` or `B` (you only get an injection from `A | B` to any other upper bound of `A` and `B`...)  So I tend to see why that error makes sense, but the fact that the behavior changes when you use `A | B` instead of an alias is a bit confusing (I think it would be reasonable to have limitations in this area).  It "feels" like when you use `t: type[A | B]` you are trying to say that `t` should be either `A`, `B`, or a subtype of either of them, and not some weird disjoint union of some of A and some of B, but this weird entity would theoretically be allowed (although I'm kind of having trouble conjuring one in python without resorting to literal / dependent types).

**Comment by erictraut** [comment]
The reason you're seeing strange behaviors is because you're creating a circularity in your type definitions. These are difficult for type checkers to detect and often benign, but they can lead to non-deterministic behaviors if the type evaluations differ depending on where in the circularity evaluation begins.

In your case, you're defining type alias `AorB` based on the decorated types `A` and `B`. The decorator uses TypeVar `TAorB` which depends on type alias `AorB`. There's no good way for a type checker to resolve this circularity. 

Pyright already has a bunch of complex logic in place for handling circularities like this, but there are limits.
